include/seqan3/range/view/pairwise_combine.hpp

```diff
@@ -0,0 +1,595 @@
+    {
+        return iterator_type{back_iterator, std::ranges::begin(src_range), std::ranges::end(src_range)};
+    }
+
+    //!\copydoc end()
+    constexpr iterator_type end() const noexcept
+    {
+        return iterator_type{back_iterator, std::ranges::begin(src_range), std::ranges::end(src_range)};
+    }
+
+    //!\copydoc end()
+    constexpr iterator_type cend() const noexcept
+    {
+        return iterator_type{back_iterator, std::ranges::begin(src_range), std::ranges::end(src_range)};
+    }
+    //!\}
+
+    /*!\name Capacity
+     * \{
+     */
+
+    /*!\brief Returns the size of the range.
+     *
+     * \details
+     *
+     * The size can only be computed when the entire range models std::ranges::RandomAccessRange.
+     *
+     * ### Complexity
+     *
+     * Constant.
+     *
+     * ### Exceptions
+     *
+     * No-throw guarantee.
+     */
+    constexpr size_t size() const noexcept
+    //!\cond
+        requires std::RandomAccessIterator<iterator_type>
+    //!\endcond
+    {
+        return static_cast<size_t>(cend() - cbegin());
+    }
+    //!\}
+
+private:
+
+    //!\brief The underling range.
+    source_range_type src_range{};
+    //!\brief The cached iterator pointing to the last element of the source range.
+    std::ranges::iterator_t<source_range_type> back_iterator{};
+};
+
+/*!\name Type deduction guides
+ * \relates seqan3::detail::pairwise_combine_view
+ * \{
+ */
+
+//!\brief Deduces the source range type from the passed view.
+template <std::ranges::View other_type>
+pairwise_combine_view(other_type view) -> pairwise_combine_view<std::remove_reference_t<other_type>>;
+
+//!\brief Deduces the source range type from the passed range.
+template <std::ranges::Range range_type>
+    requires !std::ranges::View<range_type>
+pairwise_combine_view(range_type & range) -> pairwise_combine_view<decltype(std::declval<range_type &>() | view::all)>;
+//!\}
+
+} // namespace seqan3::detail
+
+namespace seqan3::view
+{
+/*!\name General purpose views
+ * \{
+ */
+
+/*!\brief             A view adapter that generates all pairwise combinations of the elements of the source range.
+ * \tparam urng_t     The type of the range being processed. See below for requirements.
+ * \param[in] urange  The range being processed.
+ * \returns           A range over all pairwise combinations. See below for the properties of the returned range.
+ * \ingroup view
+ *
+ * \details
+ *
+ * This view generates two-element tuples representing all possible combinations of the elements of the source range
+ * while ignoring the order of the elements. If the source range has less than two elements the returned range is empty,
+ * otherwise the size of the returned view corresponds to the binomial coefficient `n choose 2`, where `n` is the
+ * size of the source range. The reference type of this range is a tuple over the reference type of the source range.
+ * Constness will not be propagated to the reference types of the source range.
```

Comment:
I think it does propagate const-ness of views (but not of non-views). This is also the behaviour of range-v3.

You are relying on `using source_iterator_type = std::ranges::iterator_t<source_range_type>;` which gives you a const_iterator if the underlying range type is const. 

Btw. is there a reason you chose not to use "underlying range type" like in all our other view and instead "source range type"?

Reply:
Const is not propagated! Calling begin on a const pairwise_combine_view does not invoke a const_iterator of this view (an iterator where the source range is explicitly marked as const). So the const qualification of the source range will not be changed. Did I misunderstand this somehow? Is this covered differently in the standard/range?
Please enlighten me.

Reply:
>Btw. is there a reason you chose not to use "underlying range type" like in all our other view and instead "source range type"?
This seems to be the typical wording in the user documentation of the ranges library.

Reply:
I did changed it back in the public API. I don't want to start a general discussion about this now.

Reply:
> Calling begin on a const pairwise_combine_view does not invoke a const_iterator of this view

I expect this will cause problems. Have you tried const-iterating your view? Does it work? You expose references to the underlying range sometimes, I expect this will fail in a const-context, because you can't expose non-const-references (which the non-const iterator creates) to a const-member.
You can have a look at other view implementations like view::take to see how this is handled. If you don't change this it will result in your view non being const-iterable unless the underlying range is by chance also of const-type (in which case it would work, because your iterator delegates to a const-iterator anyway and the exposed references would be const references).

> This seems to be the typical wording in the user documentation of the ranges library.

But it is different from the wording we agreed on for SeqAn3 very early (I remember that we had a long discussion on this). And we use "underlying range" consistently throughout all of the view documentation, the range landing page etc etc. We can discuss if we want to change it, but I would really like to keep it consistent for now...